### PR TITLE
Issue 1544 - Resources for issues

### LINKED ISSUE #1544: Attach resource
### Description

Adding the Groups feature into 3D Repo, allowing the user to create/edit/delete groups

### Goals

- [ ] Collaborators/admins can create groups
- [ ] viewers cannot create groups
- [ ] clicking on a group highlights the group objects in viewer

### Tasks

- [ ] A new Group panel
- [ ] List group component
- [ ] edit group component
- [ ] new API end point to remove groups
- [ ] New database schema for groups (present it here)

### Description
Highlighting objects on a federation node that would highlight multiple sub models is not functioning.

### Steps to replicate
- click on the root node of a federated tree
- The entire model should be highlighted, but only one of the models was highlighted.

### Bugs
- [ ] Bug 1 
- [ ] Bug 2 
- [ ] Bug 3 
 
### UI Oddities
- [ ] Weird UI 1 
- [ ] Weird UI 2 
- [ ] Weird UI 3 

https://xd.adobe.com/view/b485f5c3-1d06-4858-7439-a1f2ad1c4a5d-c2b4/

### PR DESCRIPTION
**Pull request by sanmont3drepo**
This fixes #1544

### CODE REVIEW

File: frontend/routes/components/resources/attachResourcesDialog/attachResourceFiles.component.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import * as React from 'react';
+import {ResourcesDropzone} from './attachResourcesDropzone';
+import { FieldArray, Field } from 'formik';
+import { TextField } from '@material-ui/core';
+import { RemoveButton } from '../resources.component';
+import { StyledFormControl,
+		FieldsRow } from '../../../viewer/components/risks/components/riskDetails/riskDetails.styles';
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourceFiles.component.tsx]:
```suggestion
import {
 StyledFormControl,
 FieldsRow
} from '../../../viewer/components/risks/components/riskDetails/riskDetails.styles';
```

Comment (simplypixi) [reply frontend/routes/components/resources/attachResourcesDialog/attachResourceFiles.component.tsx]: And tbh, you shouldn't use styles of a different component. The better approach is moving common style to shared space.

File: frontend/routes/components/resources/attachResourcesDialog/attachResourceFiles.component.tsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import * as React from 'react';
+import {ResourcesDropzone} from './attachResourcesDropzone';
+import { FieldArray, Field } from 'formik';
+import { TextField } from '@material-ui/core';
+import { RemoveButton } from '../resources.component';
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourceFiles.component.tsx]:
```suggestion
import { ResourcesDropzone } from './attachResourcesDropzone';
import { RemoveButton } from '../resources.component';
```

File: frontend/routes/components/resources/attachResourcesDialog/attachResourceFiles.component.tsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import { StyledFormControl,
+		FieldsRow } from '../../../viewer/components/risks/components/riskDetails/riskDetails.styles';
+import { ResourcesListContainer, ResourceListItem, ResourcesListScroller } from './attachResourcesDialog.styles';
+import * as filesize from 'filesize';
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourceFiles.component.tsx]: absolute import should be placed next to other absolute imports

File: frontend/routes/components/resources/attachResourcesDialog/attachResourceFiles.component.tsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import { get } from 'lodash';
+
+interface IProps {
+	files: any[];
+	uploadLimit: number;
+	validateQuota: (files: []) => boolean;
+	validateUploadLimit: (files: []) => boolean;
+}
+
+const extensionRe = /\.(\w+)$/;
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourceFiles.component.tsx]: It's better to move this pattern to helpers/constants.

File: frontend/routes/components/resources/attachResourcesDialog/attachResourceUrls.component.tsx [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+interface IState {
+	urls: IResourceUrl[];
+}
+
+const schema = Yup.object().shape({
+	});
+
+const LinkEntry = ({onClickRemove, index }) => {
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourceUrls.component.tsx]: It's too big to be here. It should be created as a new component.

File: frontend/routes/components/resources/attachResourcesDialog/attachResourceUrls.component.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	</FieldsRow>
+	);
+};
+
+export class AttachResourceUrls extends React.PureComponent<IProps, IState> {
+	public render() {
+		const { links } = this.props;
+
+		return (
+				<FieldArray
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourceUrls.component.tsx]: unnecessary indentation

File: frontend/routes/components/resources/attachResourcesDialog/attachResourceUrls.component.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+interface IResourceUrl {
+	name: string;
+	url: string;
+}
+
+interface IState {
+	urls: IResourceUrl[];
+}
+
+const schema = Yup.object().shape({
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourceUrls.component.tsx]: ?

File: frontend/routes/components/resources/attachResourcesDialog/attachResourceUrls.component.tsx [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import * as React from 'react';
+import { TextField, Button, Grid } from '@material-ui/core';
+import { Formik, Form, FieldArray, Field } from 'formik';
+import { DialogButtons } from './attachResourcesDialogButtons';
+import { StyledFormControl,
+	FieldsRow } from '../../../viewer/components/risks/components/riskDetails/riskDetails.styles';
+import { ResourceListLinkItem, ResourcesListContainer,
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourceUrls.component.tsx]: You mixed absolute and relative imports

File: frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.component.tsx [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import { Form, Formik } from 'formik';
+import * as React from 'react';
+import * as Yup from 'yup';
+import { AttachResourceFiles } from './attachResourceFiles.component';
+import { AttachResourceUrls } from './attachResourceUrls.component';
+import {
+		DialogTabs, DialogTab
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.component.tsx]: weird line-break

File: frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.component.tsx [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+					indicatorColor="primary"
+					textColor="primary"
+					onChange={this.handleTabChange}
+				>
+					<DialogTab label="Files" />
+					<DialogTab label="Links" />
+				</DialogTabs>
+
+				<Formik
+				validationSchema={schema}
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.component.tsx]: indentation

File: frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.component.tsx [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+				initialValues={{ files: [], links: [] }}
+				onSubmit={this.onSubmit}
+				render={({ values }) => (
+					<Form>
+						{
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.component.tsx]: Why didn't you use `renderWhenTrue`?

File: frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.styles.ts [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { DialogContent, Divider, InputAdornment, Typography } from '@material-ui/core';
+import styled from 'styled-components';
+import Dropzone from 'react-dropzone';
+import * as React from 'react';
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.styles.ts]: I don't need React import here

File: frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.styles.ts [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+	& >:first-child {
+		flex-grow: 1;
+	}
+`;
+
+export const AddLinkContainer = styled.div`
+	margin-top: 20px;
+`;
+
+export const ExtensionAdornment =  styled((props) => {
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.styles.ts]:
```suggestion
export const ExtensionAdornment =  styled(InputAdornment).attrs({
   position: 'end'
})`
 // styles
`
```

File: frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.styles.ts [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+	return React.createElement(InputAdornment,  {...props, position: 'end'});
+})`
+	p {
+		font-weight: bold;
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.styles.ts]: Could you use `FONT_WEIGHT` constant?

File: frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.styles.ts [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+		font-family: courier;
+	}
+`;
+
+export const DropzoneContent = styled.div`
+	cursor: ${(props: any) => props.error ? 'default' : 'pointer' };
+	background-color: ${(props: any) =>
+		props.error ? COLOR.WARNING_LIGHT :
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.styles.ts]: Isn't better to create some helper function?

File: frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialogButtons.tsx [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import * as React from 'react';
+import {
+		NeutralActionButton,
```
Comment (simplypixi) [inline frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialogButtons.tsx]: wrong indentation and mixed imports

File: frontend/routes/components/resources/resources.component.tsx [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import * as React from 'react';
+import { RemoveIcon, IconButton,
+		ResourceItemContainer, ResourceLink, PhotoIcon, LinkIcon,
+		DocumentIcon, ResourceLabel, UploadSizeLabel, ResourcesContainer } from './resources.styles';
```
Comment (simplypixi) [inline frontend/routes/components/resources/resources.component.tsx]: wrong indentation and mixed imports

File: frontend/routes/components/resources/resources.component.tsx [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+		aria-haspopup="true"
+	>
+		<RemoveIcon />
+	</IconButton>
+);
+
+const imageExtensions = ['png', 'jpg', 'jpeg', 'gif', 'bmp', 'tiff', 'pcx'];
+
+const ResourceIcon = ({type}) =>
+	(type === 'http') ?
```
Comment (simplypixi) [inline frontend/routes/components/resources/resources.component.tsx]: I think it will look better if you use `cond`

Comment (simplypixi) [reply frontend/routes/components/resources/resources.component.tsx]: And don't need to define the same color on each variant. Instead, you can set it on `ResourceIcon`

File: frontend/routes/components/resources/resources.styles.ts [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+`;
+
+export const UploadSizeLabel = styled.div`
+	padding-right: 28px;
+`;
+
+const Icon = (icon) => styled(icon).attrs({
+	classes: {
+		colorSecondary: 'secondary',
+		colorPrimary: 'primary',
```
Comment (simplypixi) [inline frontend/routes/components/resources/resources.styles.ts]: Is it necessary? You are using only `primary` variant, as far as I can see

File: frontend/routes/components/resources/resources.component.tsx [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+				}
+		});
+	}
+
+	public render() {
+		const resources = this.props.resources || [];
+
+		return (
+			<ResourcesContainer>
+				<FieldLabel >Resources</FieldLabel>
```
Comment (simplypixi) [inline frontend/routes/components/resources/resources.component.tsx]:
```suggestion
				<FieldLabel>Resources</FieldLabel>
```

File: frontend/routes/viewer/components/issues/components/issueDetails/issueDetails.component.tsx [hunk 21]
```diff
@@ -150,7 +155,8 @@ export class IssueDetails extends React.PureComponent<IProps, IState> {
 	}
 
 	public renderDetailsForm = () => {
-		const {issue} = this.props;
+		const {issue, onRemoveResource, showDialog, topicTypes,
+				currentUser, myJob, attachFileResources, attachLinkResources} = this.props;
```
Comment (simplypixi) [inline frontend/routes/viewer/components/issues/components/issueDetails/issueDetails.component.tsx]: weird line-break

File: frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.component.tsx (hunk 11 above)
Comment (sanmont3drepo) [reply frontend/routes/components/resources/attachResourcesDialog/attachResourcesDialog.component.tsx]: I didnt see the need for it

File: frontend/routes/components/resources/resources.component.tsx (hunk 18 above)
Comment (sanmont3drepo) [reply frontend/routes/components/resources/resources.component.tsx]: What do you mean with cond? conditional rendering or what?

File: frontend/modules/issues/issues.redux.ts [hunk 22]
```diff
@@ -171,6 +182,33 @@ const resetComponentState = (state = INITIAL_STATE) => {
...
+
+const attachResourcesSuccess = (state = INITIAL_STATE, { resources, issueId }) => {
+	resources = resources.concat(state.issuesMap[issueId].resources);
+	const issuesMap = updateIssueProps(state.issuesMap, issueId, { resources });
+	return { ...state, issuesMap};
+};
+
+const updateResourcesSuccess = (state = INITIAL_STATE, { resourcesIds, updates, issueId }) => {
+	const resources = state.issuesMap[issueId].resources.map((r) => {
+		const updateIndex = resourcesIds.indexOf(r._id);
```
[the remark this answers is not in the document]
Comment (sanmont3drepo) [reply frontend/modules/issues/issues.redux.ts]: The value updateIndex is used below

**Comment by Charence** [comment]
NOTE: fs.levels needs to be defined in config

File: backend/handler/externalServices.js [hunk 23]
```diff
@@ -21,9 +21,12 @@ const S3Handler = require("./s3");
 const GridFSHandler = require("./gridfs");
 const ResponseCodes = require("../response_codes");
 const SystemLogger = require("../logger.js").systemLogger;
+const config = require("../config.js");
 
 const ExternalServices = {};
 
+const getDefaultStorageType = () => config.defaultStorage || config.fs || "gridfs";
```
Comment (Charence) [inline backend/handler/externalServices.js]: Are we sure about `config.fs`?

**Comment by sanmont3drepo** [comment]
@Charence Fixed the system comments thing. I'll address the rest when @PavolKnapo  takes a look at your comments.
On another note Pavol told me that commenters are the ones authorised to attach resources. Maybe you two wants to review that?

**Comment by paloknapo** [comment]
@sanmont3drepo @Charence Commenter can create an issue, hence can upload or remove resources. Only viewer can't upload or remove, just view resources.

**Pull request merged**